ws/adapter.go

```diff
@@ -0,0 +1,83 @@
+package ws
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux"
+	broker "github.com/nats-io/go-nats"
+)
+
+var _ Service = (*adapterService)(nil)
+
+// Service contains publish and subscribe methods necessary for
+// message transfer.
+type Service interface {
+	mainflux.MessagePublisher
+
+	// HandleMessage that is received from message broker.
+	HandleMessage(*broker.Msg)
+
+	// AddConnection adds new client ws connection for given client and channel.
+	AddConnection(string, string, *websocket.Conn)
+}
+
+type adapterService struct {
+	pub    mainflux.MessagePublisher
+	conns  map[string]map[string]*websocket.Conn
+	logger log.Logger
+}
+
+// New instantiates the domain service implementation.
+func New(pub mainflux.MessagePublisher, logger log.Logger) Service {
+	return &adapterService{
+		pub:    pub,
+		conns:  make(map[string]map[string]*websocket.Conn),
+		logger: logger,
+	}
+}
+
+func (as *adapterService) Publish(msg mainflux.RawMessage) error {
+	return as.pub.Publish(msg)
+}
+
+func (as *adapterService) HandleMessage(msg *broker.Msg) {
```

Comment:
Based on the discussion we had above, I would advise that you make this method private and run it as part of the `Publish`, i.e. no need to publish to NATS, and then consume it back (for now).